### PR TITLE
Get soft deletes

### PR DESCRIPTION
**Pull request by mikemrm**
Adds getting of soft deleted items.

Get and List functions may now provide an additional options argument which allows the caller to provide additional settings for the requests.

The first of which is `WithDeleted` which, when set `true` will allow for deleted directories to be retrieved.

This works for both getting a directory or returning a list of directories, such as getting children or parents.

### CODE REVIEW

File: app/v1/fullsubtree_reconciler.go
Comment (JAORMX) [inline app/v1/fullsubtree_reconciler.go:37]: The new addition of options changes the function signature. While I really like that we now take options into use, having to add `nil` to the function when using default values is not great developer experience (devex). Let's use variadic parameters instead so the resulting call ends up as:

```golang
subdirs, err := c.c.GetChildren(ctx, c.baseDir)
```

File: app/v1/fullsubtree_reconciler.go
Comment (JAORMX) [inline app/v1/fullsubtree_reconciler.go:61]: Same comment, let's use variadic parameters instead.

File: client/v1/impl.go
Comment (JAORMX) [inline client/v1/impl.go:118]: You could turn this into a variadic function by changing the definition to:

```golang
func (c *httpClient) ListRoots(ctx context.Context, opts... storage.ListOptions) (*v1.DirectoryList, error) {
```

Note that I've omitted the usage of a pointer because we want to ensure that users calling this function pass valid references of `ListOptions`

**Review by JAORMX — CHANGES_REQUESTED** [review]
I like where this is going.

I did notice that we could improve devex by changing the way we're adding options to client calls though.

Instead of having an options struct and passing that around... What about using a "builder" pattern instead? Have an abstraction called `CRUDBuilder` or `URLBuilder`, You could then start adding options either from a struct or from `With*` functions. That pattern is follows throughout this code-base, so you should be able to find relevant examples.

For instance the [storage driver builder](https://github.com/infratographer/fertilesoil/blob/main/storage/crdb/driver/driver.go#L23) is a variadic function:

```golang
func NewDirectoryDriver(db *sql.DB, opts ...Options) *Driver {
```

And [options](https://github.com/infratographer/fertilesoil/blob/main/storage/crdb/driver/options.go) are simply functions that can be dynamically added to the builder call.

In our code-base, the basic call with default parameters would look as follows:

```golang
dir, err := c.c.GetDirectory(ctx, dir)
```

Adding options, say `WithDeletedDirectories` would look as follows:

```golang
dir, err := c.c.GetDirectory(ctx, dir, WithDeletedDirectories)
```

The `GetDirectory` function would need to be in charge of calling the options functions and building whatever needs to be built.

The idea is to optimize for the most common case. Since using the defaults will be the most common case, it's what should be the easiest to write.

Does this make sense?

I'd also be open to listening to counterarguments for this approach.

File: storage/crdb/driver/driver.go
Comment (JAORMX) [inline storage/crdb/driver/driver.go:144]: Thanks for removing the `%[1]s` from this query. I had missed that it wasn't appropriate to add to this specific one.

File: storage/crdb/driver/driver.go
Comment (JAORMX) [inline storage/crdb/driver/driver.go:264]: Seems we need to add tests for this case.

File: storage/crdb/driver/driver.go
Comment (JAORMX) [inline storage/crdb/driver/driver.go:73]: needs tests

**Review by JAORMX — CHANGES_REQUESTED** [review]

**Comment by mikemrm** [comment]
This does make sense. I had considered this as well, but ultimately had gone the initial way. I've now converted it to using this new method and your other issues.

**Review by fishnix — APPROVED** [review]
lots going on here, the builder pattern is much nicer. lgtm 👍

**Review by JAORMX — APPROVED** [review]
LGTM

thanks for addressing the comments. Let's wait for @fishnix to review before merging.

**Pull request merged**
